### ISSUE TITLE
Add Capability to Recommend Follow-Up Tutorials on Tutorials

### ISSUE THREAD

**Issue by sabine** [ocaml/ocaml.org#1589]
We would love to have an external contributor work on this issue.

# Task

It would be helpful to recommend to people at the end of a tutorial further tutorials that we recommend them to read after having finished the current one.

To achieve that, we need the ability for tutorials to reference the tutorials that we recommend them to read next.

1. You need to add a field `recommended_next_tutorials: string list` to both the `metadata` and `t` types in `tutorial.ml`. which holds the slugs of all the tutorials we recommend reading after this one.
2. Add some sensible references between tutorials in the `data/tutorials` folder, for testing purposes.
3. In `handler.ml`, in the function `tutorial`, you need to filter the tutorials for those that should be recommended as next reading, i.e. (`slug` is an element of `tutorial.recommended_next_tutorials`).
4. In `tutorial.eml`, render a plain list of links to the recommended tutorials at the bottom of the page.

**Important:** When a referenced tutorial slug does not exist, an error should be thrown by `ood-gen` so that we can be sure that the references between tutorials will stay in a good shape.

**Comment by iamkalio** [comment]
Hello @sabine @SaySayo , I'm interested  and would like to work on this task. Could you please assign it to me? Thank you
I've created a pull-request for the first task while its on review  I can start working on this if assigned. thanks

**Comment by sabine** [comment]
@kalio007 You are now assigned to this issue. Have fun! :slightly_smiling_face:

**Comment by sabine** [comment]
@kalio007 are you working on this issue? Please confirm if you still plan to work on this issue.

You can open a draft pull request if you need feedback / help.

**Comment by iamkalio** [comment]
@sabine I did have a couple issues, but ill create a draft pull request now. I've been going through GUI projects to understand them hence the delay

**Comment by PoorlyDefinedBehaviour** [comment]
I'll do it if no one else wants to (since I have already done another one on Dune)

**Comment by enoonan** [comment]
I started hacking on this but I think there's something I don't understand about these `@@deriving` annotations. I'm assuming it works similarly to `derive` macros in Rust but that's about as far as I go.

After step 1, the error I get on compile is `Didn't find the function for key: recommended_next_tutorials` which I can just grep for and see it comes from the `ppx_deriving_yaml` library. So I tried extracting recommended_next_tutorials into a type like so:

```
type recommended_next_tutorials = string list
[@@deriving of_yaml, show { with_path = false }]
```

But that has the same result. So there's some step I'm missing. Looking hard at this [decode function](https://github.com/ocaml/ocaml.org/blob/838bae2201cc7f71a5d8fb83560ee92210f3288f/tool/ood-gen/lib/tutorial.ml#L114). Can't tell if I'm close to cracking it or if I'm just in way over my head!

**Comment by enoonan** [comment]
Figured it out! I was missing the data.mli definition. Gonna add some sensible references in the morning and make a PR

**Comment by sabine** [comment]
Oh great!

Yes, indeed `deriving` is similar to Rust's `derive`. It's a way to run a PPX (preprocessor extension) that transforms the AST (in this case, to add functions to parse values of the type from YAML and to render values of the type to a string).

**Comment by enoonan** [comment]
Pull request here: https://github.com/ocaml/ocaml.org/pull/1928 

(I don't know how to tie a PR directly to an issue like this)

**Comment by enoonan** [comment]
> I'll do it if no one else wants to (since I have already done another one on Dune)

I am sorry I totally missed this @PoorlyDefinedBehaviour, hope I didn't stomp on you!

**Comment by PoorlyDefinedBehaviour** [comment]
> > I'll do it if no one else wants to (since I have already done another one on Dune)
> 
> I am sorry I totally missed this @PoorlyDefinedBehaviour, hope I didn't stomp on you!

Not a problem at all :+1:

**Comment by sabine** [comment]
Resolved by #1928
